src/data-loaders/navigation-guard.ts

```diff
@@ -269,6 +175,152 @@ export function isAsyncModule(
+                }
+              }
+            }
+          )
+
+          lazyLoadingPromises.push(promise)
+        }
+      }
+    }
+  }
+
+  return Promise.all(lazyLoadingPromises).then(() => {
+    // group all the loaders in a single set
+    for (const record of to.matched) {
+      // merge the whole set of loaders
+      for (const loader of record.meta[LOADER_SET_KEY]!) {
+        to.meta[LOADER_SET_KEY]!.add(loader)
+      }
+    }
+    // we return nothing to remove the value to allow the navigation
+    // same as return true
+  })
+}
+
+export function executeLoaders({
+  app,
+  loaders,
+  to,
+  router,
+  effect,
+  isSSR,
+  selectNavigationResult = (results) => results[0]!.value,
+}: {
+  app: App<unknown>
+  router: Router
+  loaders: UseDataLoader[]
+  to: RouteLocationNormalizedLoaded
+  effect: EffectScope
+  isSSR?: boolean
+  selectNavigationResult?: DataLoaderPluginOptions['selectNavigationResult']
+}) {
+  // TODO: could we benefit anywhere here from verifying the signal is aborted and not call the loaders at all
+  // if (to.meta[ABORT_CONTROLLER_KEY]!.signal.aborted) {
+  //   return to.meta[ABORT_CONTROLLER_KEY]!.signal.reason ?? false
+  // }
+
+  // unset the context so all loaders are executed as root loaders
+  setCurrentContext([])
+  return Promise.all(
+    loaders.map((loader) => {
+      const { server, lazy } = loader._.options
+      // do not run on the server if specified
+      if (!server && isSSR) {
+        return
+      }
+      // keep track of loaders that should be committed after all loaders are done
+      const ret = effect.run(() =>
+        app
+          // allows inject and provide APIs
+          .runWithContext(() =>
+            loader._.load(to as RouteLocationNormalizedLoaded, router)
+          )
+      )!
+
+      // on client-side, lazy loaders are not awaited, but on server they are
+      // we already checked for the `server` option above
+      return !isSSR && lazy
+        ? undefined
+        : // return the non-lazy loader to commit changes after all loaders are done
+          ret
+    })
+  ) // let the navigation go through by returning true or void
+    .then(() => {
+      // console.log(
+      //   `✨ Navigation results "${to.fullPath}": [${to.meta[
+      //     NAVIGATION_RESULTS_KEY
+      //   ]!.map((r) => JSON.stringify(r.value)).join(', ')}]`
+      // )
+      if (to.meta[NAVIGATION_RESULTS_KEY]!.length) {
+        return selectNavigationResult(to.meta[NAVIGATION_RESULTS_KEY]!)
+      }
+    })
+    .catch((error) =>
+      error instanceof NavigationResult
+        ? error.value
+        : // let the error propagate to router.onError()
+          // we use never because the rejection means we never resolve a value and using anything else
+          // will not be valid from the navigation guard's perspective
+          Promise.reject<never>(error)
+    )
+}
+
+export async function preloadRoute(router: Router, route: RouteLocationRaw) {
```

Comment:
I assume this function wouldn't actually want to live here eventually but I just wanted to keep it close while I started investigating for simplicity.